src/udev/udev-builtin-net_id.c

```diff
@@ -600,6 +603,93 @@ static int names_platform(sd_device *dev, NetNames *names, bool test) {
         return 0;
 }
 
+#define DEVICETREE_ALIASES "/proc/device-tree/aliases"
+
+static int dev_devicetree_onboard(sd_device *dev, NetNames *names) {
+        const char *alias, *ofnode_path, *ofnode_syspath, *parent_syspath;
+        _cleanup_free_ char *devicetree_syspath = NULL;
+        sd_device *aliases, *ofnode, *parent;
+        int r;
+        unsigned index;
+
+        if (!naming_scheme_has(NAMING_DEVICETREE_ALIASES))
+                return 0;
+
+        /* check if our direct parent has an of_node */
+        r = sd_device_get_parent(dev, &parent);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_syspath(parent, &parent_syspath);
+        if (r < 0)
+                return r;
+
+        r = sd_device_new_child(&ofnode, parent, "/of_node");
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_syspath(ofnode, &ofnode_syspath);
+        if (r < 0)
+                return r;
+
+        r = chase_symlinks("/proc/device-tree", NULL, 0, &devicetree_syspath, NULL);
+        if (r < 0)
+                return r;
+
+        /*
+         * Example paths:
+         * devicetree_syspath = /sys/firmware/devicetree/base
+         * ofnode_syspath = /sys/firmware/devicetree/base/soc/ethernet@deadbeef
+         * ofnode_path = soc/ethernet@deadbeef
+         */
+        ofnode_path = path_startswith(ofnode_syspath, devicetree_syspath);
+        if (!ofnode_path)
+                return -ENOENT;
+
+        r = sd_device_new_from_syspath(&aliases, strjoina(devicetree_syspath, "/aliases"));
+        if (r < 0)
+                return r;
+
+        FOREACH_DEVICE_SYSATTR(aliases, alias) {
+                const char *alias_path, *alias_index, *conflict = "ethernet";
+
+                alias_index = startswith(alias, "ethernet");
+                if (!alias_index)
+                        continue;
+
+                r = sd_device_get_sysattr_value(aliases, alias, &alias_path);
+                if (r < 0)
+                        continue;
+
+                /* ofnode_path lacks a leading /, so we need to adjust */
+                if (*alias_path != '/' || !streq(ofnode_path, alias_path + 1))
+                        continue;
+
+                /* If there's no index, we default to 0... */
+                if (streq(alias, "ethernet")) {
+                        conflict = "ethernet0";
+                } else {
+                        r = safe_atou(alias_index, &index);
+                        if (r < 0)
+                                return log_device_debug_errno(dev, r,
+                                        "could not get index of alias %s\n", alias);
+                }
```

Comment:
Please move the declaration of `index` in the loop. Also, you can use `alias_index` here. How about the following:
```c
const char *conflict;

...

if (isempty(alias_index)) {
        index = 0;
        conflict = "ethernet0";
} else {
        r = safe_atou(...);
        if (r < 0)
                return log_device_debug_errno(...);
        conflict = "ethernet";
}

if (index == 0 && sd_device_get_sysattr_value(aliases, conflict, NULL) >= 0)
        return log_device_debug_errno(dev, SYNTHETIC_ERRNO(EEXIST), "...");
```

Reply:
That looks good.